### ISSUE TITLE
getline() is broken

### ISSUE THREAD

**Issue by Shougo** [neovim/neovim#7859]
- `nvim --version`: NVIM v0.2.3-415-g9ea1752
- Vim (version: ) behaves differently?: Yes.  In Vim8, no errors.
- Operating system/version: Ubuntu 16.04, Manjaro Linux
- Terminal name/version: st 0.7
- `$TERM`:

### Steps to reproduce using `nvim -u NORC`

It cannot be reproduced in `nvim -u NORC`.

### Actual behaviour

```
$ build/bin/nvim codecov.yml
Error detected while processing BufReadPost Auto commands for "*":
E691: Can only compare List with List
E15: Invalid expression: getline(1) =~# '\v^#!%(\f*/|/usr/bin/env\s*<)fish>' |     setlocal filetype=fish | endif
Error detected while processing function <SNR>11_DetectJS:
line    1:
E691: Can only compare List with List
E15: Invalid expression: getline(1) =~# '^#!.*/bin/\(env\s\+\)\?node\>'
Press ENTER or type command to continue
```

`getline(0)` returns `[]` when file opening.
It breaks the plugins.

### Expected behaviour

No errors.

### `git bisect` result

I have bisected the problem.

9ea1752d60589e8fc5e7184144bc6d1c1b9f16a3 is the first bad commit
commit 9ea1752d60589e8fc5e7184144bc6d1c1b9f16a3
Author: ZyX <kp-pav@yandex.ru>
Date:   Wed Jan 3 00:00:16 2018 +0300

    *: Provide list length when allocating lists

:040000 040000 3e7b3a835596a1326b8c8539cdb650ee7fa32c18 1f0b5f46580434e9a4d27a82e57929b06792f281 M      src
:040000 040000 68c54453675b5b399c9339ab4b93a1ad686b6c28 6fd87484a6feb623f04ee1f3404da34f192ab5d4 M      test

**Comment by Shougo** [comment]
In neovim, `getline(-1)` returns `[]`.
In Vim8, `getline(-1)` returns `''`.

It is incompatible behavior.

**Comment by Shougo** [comment]
It is fixed by #7860.
